graphql_compiler/compiler/match_query.py

```diff
@@ -132,6 +134,49 @@ def _split_match_steps_into_match_traversals(match_steps):
     return output
 
 
+def _extract_folds_from_ir_blocks(ir_blocks):
+    """Extract all @fold data from the IR blocks, and cut the folded IR blocks out of the IR.
+
+    Args:
+        ir_blocks: list of IR blocks to extract fold data from
+
+    Returns:
+        tuple (folds, remaining_ir_blocks):
+        - folds: dict of FoldScopeLocation -> list of IR blocks corresponding to that @fold scope.
+                 The list is guaranteed to start with a Fold and end with an Unfold block.
+        - remaining_ir_blocks: list of IR blocks that were not part of a Fold-Unfold section.
+    """
+    folds = dict()
+    remaining_ir_blocks = []
+    current_folded_blocks = []
+    in_fold = False
+
+    for block in ir_blocks:
+        if isinstance(block, Fold):
+            in_fold = True
+
+        if in_fold:
+            current_folded_blocks.append(block)
+        else:
+            remaining_ir_blocks.append(block)
+
+        if isinstance(block, Unfold):
+            in_fold = False
+            first_folded_block = current_folded_blocks[0]
+            last_folded_block = current_folded_blocks[-1]
+
+            if not isinstance(first_folded_block, Fold):
```

Comment:
Shouldn't this not happen because of the way the scoping for the fold works ?

Reply:
Indeed it should not. Whenever you see `raise AssertionError`, that's my way of saying "this should never ever happen."